### PR TITLE
Add output of group and field phase potential injection and production rates to Summary file

### PR DESCRIPTION
**Pull request by jalvestad**
This pullrequest contains an extension of the output of phase potential injection and production rates to Summary file for group levels and field level. The work here is based on the same output for wells, done in pullrequest #554 by @totto82 who also is informed about this addition. 

I will thank my collaborator @bska for valuable discussions and advice during the work.

### CODE REVIEW

**Comment by bska** [comment]
Thank you&mdash;this is a very welcome extension and I think you've done all you need to do in the opm-common module.  I do see one possible limitation in the simulator code of opm-simulators though.

According to [`BlackoilWellModel<>::computeWellPotentials()`](https://github.com/OPM/opm-simulators/blob/51d85d25d103ccf945b2a624fb3ff22544c3b892/opm/autodiff/BlackoilWellModel_impl.hpp#L974-L979) the potential values will only be calculated and stored in the `Rates` object **if** the pertinent well is associated to one of well-related potential keywords (e.g., `WOPP` or `WGIP`).  That restriction is probably too limited in the context of field- and group-level potential rates output to the summary file.

I have not looked too closely at the implementation in this proposal so I invite other interested parties to have a look and to make other comments.

**Comment by bska** [comment]
jenkins build this please

**Comment by bska** [comment]
jenkins build this please

**Comment by bska** [comment]
Thanks for the updates in 244522.  As far as I can tell the remaining test failures are all of the form "new run outputs additional summary vectors" which we can incorporate by `update_data`.  Unless anyone objects I intend to update the reference data and install this into master no later than about noon CET tomorrow (Friday 14th December).

**Comment by joakim-hove** [comment]
> As far as I can tell the remaining test failures are all of the form "new run outputs additional summary vectors" which we can incorporate by update_data. 

Yes - that seems correct; but I really think also the summary comparison should respect the `ignore_extra` setting.


> Unless anyone objects I intend to update the reference data and install this into master [...]

I had one very minor comment about the name `generic_potential_rate` which I did not really understand; other than that no comments.

**Comment by bska** [comment]
> I really think also the summary comparison should respect the `ignore_extra` setting.

That's a good point.  I have a work-in-progress PR (OPM/opm-simulators#1673) that I'm using for the restart work.  I'll update the comparison scripts to make the summary comparisons aware of `ignore_extra` too and then I'll rerun Jenkins on this PR.

**Comment by joakim-hove** [comment]
>I'll update the comparison scripts to make the summary comparisons aware of ignore_extra too

Unfortunately I think it is more work than that - the actual comparison application must be updated.

**Comment by bska** [comment]
With #587 merged, I launched [a test](https://ci.opm-project.org/job/opm-simulators-PR-builder/1251/) from [OPM/opm-simulators#1637](https://github.com/OPM/opm-simulators/pull/1673#issuecomment-447059562) using `ignore_extra` that came back green.  Once we resolve the function naming matter for `generic_potential_rate()`, I think we can safely run `update_data` and bring this into master.

File: src/opm/output/eclipse/Summary.cpp [hunk 1]
```diff
@@ -656,19 +656,24 @@ quantity region_rate( const fn_args& args ) {
 }
 
 template < rt phase, bool outputProducer = true, bool outputInjector = true>
-quantity generic_well_rate (const fn_args& args  ) {
-    const quantity zero = { 0, rate_unit< phase >() };
-    if( args.schedule_wells.empty() ) return zero;
-
-    const auto p = args.wells.find( args.schedule_wells.front()->name() );
-    if( p == args.wells.end() ) return zero;
+inline quantity generic_potential_rate( const fn_args& args ) {
```
[the remark this answers is not in the document]
Comment (joakim-hove) [reply src/opm/output/eclipse/Summary.cpp]: Seems my followup to this was lost in the internet; anyway - my main point was not about the `generic` - but rather about the `potential` - which I frankly do not understand. The way I understand this function `generic_rate` would be a good name?

[the remark this answers is not in the document]
Comment (GitPaean) [reply src/opm/output/eclipse/Summary.cpp]: Then it is difficult to tell it is about `potential` now.  `potential` is one type of `rate`, but it is a very `specific` type of `rate`, not the `rate` we usually refer to, like `WOPR`

**Comment by joakim-hove** [comment]
OK:

1. I am ready to be merge this; however as I have indicated I would prefer e.g. the name `generic_rate` instead of the current `generic_potential_rate`. The reason I am slightly insistent on this is partly due to an inkling feeling that maybe I actually misunderstand this - if you feel `generic_rate` is just plain wrong; that might be because I misunderstand something. 

2. As I understand it this can be merged - and the `update_data` performed - without OPM/opm-simulators#1637 - if that is correct I think we should do it like that. If it is not correct - the dependency should be visible from the PR description. I'll fire up a jenkins run now.

**Comment by joakim-hove** [comment]
jenkins build this ignore_extra please

File: src/opm/output/eclipse/Summary.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (joakim-hove) [reply src/opm/output/eclipse/Summary.cpp]: > Then it is difficult to tell it is about potential now. potential is one type of rate, but it is a very specific type of rate, not the rate we usually refer to, like WOPR

OK - I also had a feeling it was something like this; but the implementation seems to just sum up rates?

**Comment by bska** [comment]
> ignore_extra please

I seem to recall that `ignore_extra` is recognised only in the context of an opm-simulators PR.

> As I understand it this can be merged - and the `update_data` performed - without [OPM/opm-simulators#1673](https://github.com/OPM/opm-simulators/pull/1673) - if that is correct I think we should do it like that.

That is correct.  I abused OPM/opm-simulators#1673 only as a means to running this PR using `ignore_extra`.  This PR is entirely independent of OPM/opm-simulators#1673.

**Comment by jalvestad** [comment]
I have changed the function name from generic_potential_rate to potential_rate to avoid confusion about the meaning of the function name. 

Else I hope this pull request can be merged now/soon unless there are other issues remaining.

**Comment by joakim-hove** [comment]
> Else I hope this pull request can be merged now/soon unless there are other issues remaining.

Sure!

**Comment by bska** [comment]
> Cool if you rebase it on latest master.

I don't quite understand that comment.  The parent of 3429db8 is current master (fc0e61518).  Am I missing something?

**Comment by joakim-hove** [comment]
Another `test_update` sneaked by - will play the `update_data` dance one more time.

**Comment by joakim-hove** [comment]
jenkins build this update_data please

**Comment by joakim-hove** [comment]
jenkins build this opm-tests=88 please

**Comment by joakim-hove** [comment]
jenkins build this opm-tests=88 please

**Comment by joakim-hove** [comment]
I am ready to merge this, but it seems unwilling to become green, even after `update_data`?

**Comment by bska** [comment]
> it seems reluctant to become green, even after `update_data`?

The last time I encountered a similar symptom, closing the PR (in this case #585) and leaving it closed for a few hours and then reopening it and running `update_data` was enough to make it go green.  I don't know if that was pure luck though...

**Comment by joakim-hove** [comment]
jenkins build this update_data please

**Comment by bska** [comment]
jenkins build this opm-tests=88 please

**Comment by bska** [comment]
Last remaining integration test failure below.  Why was the Fetkovic example not updated?

---
```
Comparing '/var/lib/jenkins/post-builder/workspace/opm-common-PR-builder/mpi/build-opm-simulators/tests/results/flow+fetkovich_2d/2D_FETKOVICHAQUIFER' to '/var/lib/jenkins/post-builder/workspace/opm-common-PR-builder/deps/opm-tests/aquifer-fetkovich/opm-simulation-reference/flow/2D_FETKOVICHAQUIFER'.
Comparing 63 steps.
Keywords missing from one case: 
FGPI FGPP FOPI FOPP FWPI FWPP GGPI:I GGPI:P 
GGPP:I GGPP:P GOPI:I GOPI:P GOPP:I GOPP:P GWPI:I GWPI:P 
GWPP:I GWPP:P WGPI:INJE01 WGPI:PROD01 WGPP:INJE01 WGPP:PROD01 WOPI:INJE01 WOPI:PROD01 
WOPP:INJE01 WOPP:PROD01 WWPI:INJE01 WWPI:PROD01 WWPP:INJE01 WWPP:PROD01 
Program threw an exception: [/var/lib/jenkins/post-builder/workspace/opm-common-PR-builder/examples/test_util/summaryRegressionTest.cpp:49] Different amount of keywords in the two summary files.
Comparing '/var/lib/jenkins/post-builder/workspace/opm-common-PR-builder/mpi/build-opm-simulators/tests/results/flow+fetkovich_2d/2D_FETKOVICHAQUIFER' to '/var/lib/jenkins/post-builder/workspace/opm-common-PR-builder/deps/opm-tests/aquifer-fetkovich/opm-simulation-reference/flow/2D_FETKOVICHAQUIFER'.
Comparing 63 steps.
Keywords missing from one case: 
FGPI FGPP FOPI FOPP FWPI FWPP GGPI:I GGPI:P 
GGPP:I GGPP:P GOPI:I GOPI:P GOPP:I GOPP:P GWPI:I GWPI:P 
GWPP:I GWPP:P WGPI:INJE01 WGPI:PROD01 WGPP:INJE01 WGPP:PROD01 WOPI:INJE01 WOPI:PROD01 
WOPP:INJE01 WOPP:PROD01 WWPI:INJE01 WWPI:PROD01 WWPP:INJE01 WWPP:PROD01 
Program threw an exception: [/var/lib/jenkins/post-builder/workspace/opm-common-PR-builder/examples/test_util/summaryRegressionTest.cpp:49] Different amount of keywords in the two summary files.
```

**Comment by GitPaean** [comment]
> Why was the Fetkovic example not updated?

We did not know why `update_data` does not work for that example. So I created the PR  OPM/opm-tests#87 manually.  Maybe that was the mistake. 

I did not spot any problems in the PR https://github.com/OPM/opm-simulators/pull/1642 that causes the `update_data` not work with that example.

**Comment by akva2** [comment]
jenkins build this update_data opm-simulators=1711 please

**Comment by akva2** [comment]
jenkins build this update_data please

**Comment by bska** [comment]
jenkins build this opm-tests=88 please

**Comment by bska** [comment]
> jenkins build this opm-tests=88 please

:+1: 

Jenkins approves if we use the latest version of OPM/opm-tests#88.  I think we're cleared to merge.  If you'd like to rebase this on top of current master then feel free.  Otherwise, I think this should be merged in its current state.

**Pull request merged**